File: apps/desktop/halo_desktop/completion.py
from __future__ import annotations

from collections.abc import Iterable, Sequence

from .models import CompletionCandidate, SlashCommand


PREFIX_SCORE = 40
FUZZY_SCORE = 20
CURRENT_AGENT_SCORE = 20
RECENT_SCORE = 10
FAVORITE_SCORE = 10


def default_commands() -> tuple[SlashCommand, ...]:
    return (
        SlashCommand(
            name="/codex",
            description="Run Codex CLI in the current workspace",
            agent_id="codex-cli",
            arguments=("--continue", "--model", "--sandbox", "--full-auto"),
        ),
        SlashCommand(
            name="/review",
            description="Ask Codex CLI to review the current changes",
            agent_id="codex-cli",
            arguments=("--branch", "--changes", "--summary"),
        ),
        SlashCommand(
            name="/claude",
            description="Start a Claude Code session",
            agent_id="claude-code",
            arguments=("--continue", "--model", "--permission-mode"),
        ),
        SlashCommand(
            name="/opencode",
            description="Open an OpenCode workflow",
            agent_id="opencode",
            arguments=("--session", "--model"),
        ),
        SlashCommand(
            name="/pi",
            description="Start a Pi assistant workflow",
            agent_id="pi",
            arguments=("--persona", "--brief"),
        ),
        SlashCommand(
            name="/plan",
            description="Draft a multi-agent workflow plan",
            agent_id="codex-cli",
            arguments=("--tasks", "--dry-run"),
        ),
    )


def complete_commands(
    query: str,
    commands: Sequence[SlashCommand] | None = None,
    *,
    current_agent_id: str | None = None,
    recent: Iterable[str] = (),
    favorites: Iterable[str] = (),
) -> list[CompletionCandidate]:
    catalog = tuple(commands or default_commands())
    normalized_query = query.lstrip()
    recent_set = set(recent)
    favorite_set = set(favorites)

    first_space = _first_whitespace(normalized_query)
    if first_space >= 0:
        command_name = normalized_query[:first_space]
        argument_query = normalized_query[first_space + 1:]
        command = next(
            (item for item in catalog if item.name == command_name),
            None,
        )
        if command is not None:
            return _complete_arguments(
                command,
                argument_query.strip(),
                current_agent_id=current_agent_id,
                recent=recent_set,
                favorites=favorite_set,
            )

    candidates: list[CompletionCandidate] = []
    for command in catalog:
        score = _score_command(
            normalized_query,
            command,
            current_agent_id=current_agent_id,
            recent=recent_set,
            favorites=favorite_set,
        )
        if score > 0:
            candidates.append(
                CompletionCandidate(
                    name=command.name,
                    score=score,
                    description=command.description,
                    agent_id=command.agent_id,
                    kind="command",
                    insert_text=command.name,
                    command_name=command.name,
                )
            )

    return sorted(candidates, key=lambda item: (-item.score, item.name))


def _first_whitespace(value: str) -> int:
    for index, character in enumerate(value):
        if character.isspace():
            return index
    return -1


def _complete_arguments(
    command: SlashCommand,
    argument_query: str,
    *,
    current_agent_id: str | None,
    recent: set[str],
    favorites: set[str],
) -> list[CompletionCandidate]:
    candidates: list[CompletionCandidate] = []
    for argument in command.arguments:
        match_score = _score_text(argument_query, argument)
        if match_score == 0:
            continue

        score = match_score
        if command.agent_id == current_agent_id:
            score += CURRENT_AGENT_SCORE
        if argument in recent or f"{command.name} {argument}" in recent:
            score += RECENT_SCORE
        if argument in favorites or f"{command.name} {argument}" in favorites:
            score += FAVORITE_SCORE

        candidates.append(
            CompletionCandidate(
                name=argument,
                score=score,
                description=f"{command.name} option",
                agent_id=command.agent_id,
                kind="argument",
                insert_text=argument,
                command_name=command.name,
            )
        )
    return sorted(candidates, key=lambda item: (-item.score, item.name))


def _score_command(
    query: str,
    command: SlashCommand,
    *,
    current_agent_id: str | None,
    recent: set[str],
    favorites: set[str],
) -> int:
    match_score = _score_text(query, command.name)
    if match_score == 0:
        return 0

    score = match_score
    if command.agent_id == current_agent_id:
        score += CURRENT_AGENT_SCORE
    if command.name in recent:
        score += RECENT_SCORE
    if command.name in favorites:
        score += FAVORITE_SCORE
    return score


def _score_text(query: str, value: str) -> int:
    if query in ("", "/"):
        return FUZZY_SCORE

    query_lower = query.lower()
    value_lower = value.lower()
    if value_lower.startswith(query_lower):
        return PREFIX_SCORE
    if _is_fuzzy_continuity(query_lower, value_lower):
        return FUZZY_SCORE
    return 0


def _is_fuzzy_continuity(query: str, value: str) -> bool:
    if not query:
        return True

    search_from = 0
    for character in query:
        found_at = value.find(character, search_from)
        if found_at < 0:
            return False
        search_from = found_at + 1
    return True

File: crates/halo-core/src/completion.rs
use halo_protocol::{CompletionCandidate, SlashCommand, WorkflowKind};

const PREFIX_SCORE: u32 = 40;
const FUZZY_SCORE: u32 = 20;
const CURRENT_AGENT_SCORE: u32 = 20;
const RECENT_SCORE: u32 = 10;
const FAVORITE_SCORE: u32 = 10;

pub fn complete_commands(
    commands: &[SlashCommand],
    input: &str,
    current_agent_id: Option<&str>,
    recent_usage: &[&str],
    favorites: &[&str],
) -> Vec<CompletionCandidate> {
    if let Some((command, argument_query)) = argument_context(commands, input) {
        return complete_arguments(command, argument_query);
    }

    let query = input.trim();
    let mut candidates: Vec<_> = commands
        .iter()
        .filter_map(|command| {
            score_command(command, query, current_agent_id, recent_usage, favorites).map(|score| {
                CompletionCandidate::new(
                    command.name.clone(),
                    command.description.clone(),
                    score,
                    command.agent_id.clone(),
                )
            })
        })
        .collect();

    sort_candidates(&mut candidates);
    candidates
}

pub fn default_commands() -> Vec<SlashCommand> {
    vec![
        SlashCommand::new(
            "/codex",
            "Start a Codex CLI coding run",
            Some("codex-cli"),
            WorkflowKind::Chat,
            vec![
                "--continue".to_string(),
                "--model".to_string(),
                "--sandbox".to_string(),
            ],
        ),
        SlashCommand::new(
            "/continue",
            "Continue the active agent run",
            Some("codex-cli"),
            WorkflowKind::Chat,
            vec!["--last".to_string()],
        ),
        SlashCommand::new(
            "/claude",
            "Start a Claude Code run",
            Some("claude-code"),
            WorkflowKind::Chat,
            vec!["--resume".to_string(), "--model".to_string()],
        ),
        SlashCommand::new(
            "/opencode",
            "Start an OpenCode run",
            Some("opencode"),
            WorkflowKind::Chat,
            vec!["--model".to_string()],
        ),
        SlashCommand::new(
            "/pi",
            "Ask Pi for conversational help",
            Some("pi"),
            WorkflowKind::Chat,
            Vec::new(),
        ),
        SlashCommand::new(
            "/review",
            "Run a focused code review workflow",
            None::<String>,
            WorkflowKind::Review,
            vec!["--scope".to_string()],
        ),
        SlashCommand::new(
            "/plan",
            "Create an implementation plan",
            None::<String>,
            WorkflowKind::Planning,
            vec!["--write".to_string()],
        ),
    ]
}

fn argument_context<'a>(
    commands: &'a [SlashCommand],
    input: &'a str,
) -> Option<(&'a SlashCommand, &'a str)> {
    let trimmed = input.trim_start();
    let split_at = trimmed.find(char::is_whitespace)?;
    let command_name = &trimmed[..split_at];
    let argument_query = trimmed[split_at..].trim_start();
    let command = commands.iter().find(|item| item.name == command_name)?;

    Some((command, argument_query))
}

fn complete_arguments(command: &SlashCommand, query: &str) -> Vec<CompletionCandidate> {
    let mut candidates: Vec<_> = command
        .arguments
        .iter()
        .filter_map(|argument| {
            score_text(argument, query).map(|score| {
                CompletionCandidate::new(
                    argument.clone(),
                    format!("Argument for {}", command.name),
                    score,
                    command.agent_id.clone(),
                )
            })
        })
        .collect();

    sort_candidates(&mut candidates);
    candidates
}

fn score_command(
    command: &SlashCommand,
    query: &str,
    current_agent_id: Option<&str>,
    recent_usage: &[&str],
    favorites: &[&str],
) -> Option<u32> {
    let mut score = score_text(&command.name, query)?;

    if current_agent_id.is_some_and(|agent_id| command.agent_id.as_deref() == Some(agent_id)) {
        score += CURRENT_AGENT_SCORE;
    }

    if recent_usage.iter().any(|name| *name == command.name) {
        score += RECENT_SCORE;
    }

    if favorites.iter().any(|name| *name == command.name) {
        score += FAVORITE_SCORE;
    }

    Some(score)
}

fn score_text(candidate: &str, query: &str) -> Option<u32> {
    if query.is_empty() {
        return Some(0);
    }

    let candidate = candidate.to_ascii_lowercase();
    let query = query.to_ascii_lowercase();

    if candidate.starts_with(&query) {
        return Some(PREFIX_SCORE);
    }

    if fuzzy_contiguous_match(&candidate, &query) {
        return Some(FUZZY_SCORE);
    }

    None
}

fn fuzzy_contiguous_match(candidate: &str, query: &str) -> bool {
    let mut query_chars = query.chars();
    let Some(mut expected) = query_chars.next() else {
        return true;
    };

    for candidate_char in candidate.chars() {
        if candidate_char == expected {
            match query_chars.next() {
                Some(next) => expected = next,
                None => return true,
            }
        }
    }

    false
}

fn sort_candidates(candidates: &mut [CompletionCandidate]) {
    candidates.sort_by(|left, right| {
        right
            .score
            .cmp(&left.score)
            .then_with(|| left.name.cmp(&right.name))
    });
}
